File: enode/src/main/java/org/enodeframework/commanding/ICommandContext.java
package org.enodeframework.commanding;

import org.enodeframework.domain.IAggregateRoot;
import org.enodeframework.messaging.IApplicationMessage;

import java.util.concurrent.CompletableFuture;

public interface ICommandContext {
    /**
     * Add a new aggregate into the current command context.
     */
    void add(IAggregateRoot aggregateRoot);

    /**
     * Add a new aggregate into the current command context synchronously, and then return a completed task object.
     */
    CompletableFuture<Void> addAsync(IAggregateRoot aggregateRoot);

    /**
     * Get an aggregate from the current command context.
     */
    <T extends IAggregateRoot> CompletableFuture<T> getAsync(Object id, boolean firstFromCache, Class<T> clazz);

    <T extends IAggregateRoot> CompletableFuture<T> getAsync(Object id, Class<T> clazz);

    String getResult();

    void setResult(String result);

    /**
     * Get an application message.
     */
    IApplicationMessage getApplicationMessage();

    /**
     * Set an application message.
     */
    void setApplicationMessage(IApplicationMessage applicationMessage);

}

File: enode/src/main/java/org/enodeframework/commanding/ICommandContext.kt
package org.enodeframework.commanding

import org.enodeframework.domain.IAggregateRoot
import org.enodeframework.messaging.IApplicationMessage
import java.util.concurrent.CompletableFuture

interface ICommandContext {
    /**
     * Add a new aggregate into the current command context.
     */
    fun add(aggregateRoot: IAggregateRoot)

    /**
     * Add a new aggregate into the current command context synchronously, and then return a completed task object.
     */
    fun addAsync(aggregateRoot: IAggregateRoot): CompletableFuture<Boolean>

    /**
     * Get an aggregate sync from the current command context.
     */
    fun <T : IAggregateRoot> get(id: Any, firstFromCache: Boolean, aggregateRootType: Class<T>): T

    /**
     * Get an aggregate async from the current command context.
     */
    fun <T : IAggregateRoot> getAsync(id: Any, firstFromCache: Boolean, aggregateRootType: Class<T>): CompletableFuture<T>

    /**
     * Get an aggregate sync from the current command context, default from cache.
     */
    fun <T : IAggregateRoot> get(id: Any, aggregateRootType: Class<T>): T

    /**
     * Get an aggregate async from the current command context, default from cache.
     */
    fun <T : IAggregateRoot> getAsync(id: Any, aggregateRootType: Class<T>): CompletableFuture<T>

    /**
     * Get result.
     */
    var result: String

    /**
     * Set an application message.
     */
    var applicationMessage: IApplicationMessage?
}